[DOCS] Edits machine learning settings (#69947)
elastic/elasticsearch: docs/reference/settings/ml-settings.asciidoc

processor attempts to load the model from cache. If the `inference` processor
does not receive any documents for the duration of the TTL, the referenced model
is flagged for eviction from the cache. If a document is processed later, the

> model is again loaded into the cache. To update this setting in {ess}, see 

Author: > xpack.ml.inference_model.time_to_live

This description mentions only inference processors.  Does it also affect inference aggregation?

Reviewer: Yes inference aggregation is a user of this cache so the same setting applies. It's better not to describe the caching action in terms of the inference processor receiving documents as there could be other users of the cache.  Can you re-phase it something like this pls:

"
The time to live (TTL) for models in the inference model cache. The TTL is calculated from last access. 
Users of the cache such as the `inference` processor or the `inference` aggregator cause a model to 
be cached on first use and reset the TTL on every use. If a cached model is not accessed for the duration 
of the TTL, the referenced model is flagged for eviction from the cache. 
If a document is processed later, the model is again loaded into the cache. Defaults to `5m`.
"

Author: Thanks!  I updated this text in https://github.com/elastic/elasticsearch/pull/69947/commits/d04db2f4df24e34e4f56b66b448625b7e8667cb0. I also added the "trained model" terminology, since that's the term we use elsewhere in the docs.

`xpack.ml.max_machine_memory_percent`::
(<<cluster-update-settings,Dynamic>>) The maximum percentage of the machine's
memory that {ml} may use for running analytics processes. These processes are
separate to the {es} JVM. The limit is based on the total memory of the machine, 
not current free memory. Jobs are not allocated to a node if doing so would
cause the estimated memory use of {ml} jobs to exceed the limit. When the
{operator-feature} is enabled, this setting can be updated only by operator

> users. The minimum value is `5`; the maximum value is `200`. Defaults to `30`.

Reviewer 2: We should probably say something about the default being greater than 100 now that we are publicising it.  Something along the lines of this:

> The maximum of `200` is for the special case where it has been determined that using swap space for ML jobs is acceptable.  Do not configure a value higher than the amount of memory left over after running the Elasticsearch JVM unless you have enough swap space configured to accommodate it, and have determined that this is an appropriate configuration for a specialist use case. General best practice is not to use swap on Elasticsearch nodes, so there needs to be a well thought through reason why configuring ML to use it is a good idea in some special situation.

Author: Thanks!  I've added this as a tip in https://github.com/elastic/elasticsearch/pull/69947/commits/f9a872480f64391e589c70e765d9ff2bb9cb5041

`xpack.ml.max_model_memory_limit`::
(<<cluster-update-settings,Dynamic>>) The maximum `model_memory_limit` property
value that can be set for any job on this node. If you try to create a job with
a `model_memory_limit` property value that is greater than this setting value,
an error occurs. Existing jobs are not affected when you update this setting.

> If this setting is `0` or unset, the maximum `model_memory_limit` value is `1gb`.

Reviewer 2: ```suggestion
If this setting is `0` or unset, there is no maximum `model_memory_limit` (which makes it possible to create jobs that cannot be assigned to any available nodes).
```

Reviewer: This is true when job is created with `AnalysisLimits. model_memory_limit` set. If unset then then `AnalysisLimits. model_memory_limit` takes on the value of `xpack.ml.max_model_memory_limit` if set else 1GB. 

Suggestion:
> If this setting is `0` or unset, there is no maximum to the `model_memory_limit` setting for AD and DFA jobs (which makes it possible to create jobs that cannot be assigned to any available nodes).

Then in the [PUT Job docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/ml-put-job.html) section about `model_memory_limit` we could change 

 > The default value for jobs created in version 6.1 and later is 1024mb.

to 

> The default value for jobs created in version 6.1 and later is `xpack.ml.max_model_memory_limit`. If `xpack.ml.max_model_memory_limit` is `0` or unset `model_memory_limit` is set to `1024mb`.

Author: > ... which makes it possible to create jobs that cannot be assigned to any available nodes.

Since the first sentence of this setting description says it affects "... any job on this node", should the new clarification also be limited to this node? For example: "...which makes it possible to create jobs that cannot run on this node if it doesn't meet those memory requirements".?

small amount of time in this state before they move to `open` state. Jobs that
must restore large models when they are opening spend more time in the `opening`
state. When the {operator-feature} is enabled, this setting can be updated only 

> by operator users. The minimum value is `0`. Defaults to `2`.

Reviewer 2: I think it's a bug that the minimum is 0 with no special meaning.  I opened #70064 to discuss what to do about this.

If we leave the default as 0 with no special meaning then I think it needs an extra sentence here to explicitly call out the effect of setting 0.

small amount of time in this state before they move to `open` state. Jobs that
must restore large models when they are opening spend more time in the `opening`
state. When the {operator-feature} is enabled, this setting can be updated only 

> by operator users. The minimum value is `0`. Defaults to `2`.

Reviewer 2: ```suggestion
by operator users. Defaults to `2`.
```

I think you should temporarily remove the minimum value of 0.  Then this PR can be merged and when we fix #70064 we can add a sentence here about what 0 means after the fix.

Author: Thanks, that sounds like a good plan to me!

`xpack.ml.max_anomaly_records`::
(<<cluster-update-settings,Dynamic>>) The maximum number of records that are
output per bucket. Defaults to `500`.

`xpack.ml.max_lazy_ml_nodes`::
(<<cluster-update-settings,Dynamic>>) The number of lazily spun up {ml} nodes.
Useful in situations where {ml} nodes are not desired until the first {ml} job
opens. If the current number of {ml} nodes is greater than or equal to this 
setting, it is assumed that there are no more lazy nodes available as the
desired number of nodes have already been provisioned. If a job is opened and 
this setting has a value greater than zero and there are no nodes that can
accept the job, the job stays in the `OPENING` state until a new {ml} node is 
added to the cluster and the job is assigned to run on that node. When the
{operator-feature} is enabled, this setting can be updated only by operator

> users. The maximum value is `3`. Defaults to `0`.

Reviewer 2: There is no longer a maximum of 3 in 7.11 and above.  This was changed to support autoscaling.

```suggestion
users. Defaults to `0`.
```

`xpack.ml.max_ml_node_size`::
(<<cluster-update-settings,Dynamic>>)
The maximum node size for {ml} nodes in a deployment that supports automatic
cluster scaling. If you set it to the maximum possible size of future {ml} nodes,
when a {ml} job is assigned to a lazy node it can check (and fail quickly) when
scaling cannot support the size of the job. When the {operator-feature} is
enabled, this setting can be updated only by operator users. Defaults to `0b`,

> which means this value is ignored. 

Reviewer 2: ```suggestion
which means it will be assumed that automatic cluster scaling can add arbitrarily large nodes to the cluster. 
```
